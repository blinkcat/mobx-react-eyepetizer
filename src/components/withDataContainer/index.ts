export { withDataContainer as default } from './withDataContainer';
