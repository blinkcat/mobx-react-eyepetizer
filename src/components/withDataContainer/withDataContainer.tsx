import * as React from 'react';
import shallowEqual from 'fbjs/lib/shallowEqual';
import hoistNonReactStatic from 'hoist-non-react-statics';
import pick from 'lodash.pick';
import set from 'lodash.set';

export interface Options {
  errorHandler?: (err: { message?: string }) => JSX.Element;
  loadingHandler?: () => JSX.Element;
  pure?: boolean;
  env?: object;
  propsToWatch?: string[];
  shouldUpdate?: (prevProps, nextProps) => boolean;
  shouldSubscribe?: (prevProps, nextProps) => boolean;
}

export function withDataContainer(dataLoader, options?: Options) {
  return WrappedComponent => {
    const {
      errorHandler = err => {
        throw err;
      },
      loadingHandler = () => null,
      pure = false,
      env = {},
      propsToWatch = null,
      shouldUpdate = null,
      shouldSubscribe = null
    } =
      options || {};

    class WithDataContainer extends React.Component<any, { error: object; data: object }> {
      public target: HTMLElement;
      private _privateProps;
      private _privateCachedWatchingProps;
      private _privateUnMounted = false;

      constructor(props) {
        super(props);
        this.state = { error: null, data: null };
      }

      public componentWillReceiveProps(props) {
        this._privateSubscribe(props);
      }

      public shouldComponentUpdate(nextProps, nextState) {
        if (typeof shouldUpdate === 'function') {
          return shouldUpdate(this.props, nextProps);
        }
        if (!pure) {
          return true;
        }
        return (
          !shallowEqual(this.props, nextProps) ||
          !shallowEqual(this.state.data, nextState.data) ||
          this.state.error !== nextState.error
        );
      }

      public componentDidMount() {
        this._privateSubscribe(this.props);
      }

      public componentWillUnmount() {
        this._privateUnMounted = true;
      }

      public render() {
        const { data, error } = this.state;
        if (error) {
          return errorHandler(error);
        }
        if (!data) {
          return loadingHandler();
        }
        const finalProps = {
          ...this.props,
          ...data
        };
        const setRef = c => {
          this.target = c;
        };
        return <WrappedComponent targetRef={setRef} {...finalProps} />;
      }

      // 在这里调用dataLoader
      private _privateSubscribe(props) {
        if (this._privateShouldSubscribe(props)) {
          // update
          const onData = (error: object | null, data) => {
            if (!this._privateUnMounted) {
              this.setState({ error, data });
            }
          };
          dataLoader(props, onData, env);
        }
      }

      // 判断是否需要subscribe
      private _privateShouldSubscribe(props): boolean {
        const firstRun = !this._privateCachedWatchingProps;
        const nextProps = pick(props, propsToWatch);
        const currentProps = this._privateCachedWatchingProps || {};
        this._privateCachedWatchingProps = nextProps;
        // 第一次要重新取数据
        if (firstRun) {
          return true;
        }
        if (typeof shouldSubscribe === 'function') {
          return shouldSubscribe(currentProps, nextProps);
        }
        if (propsToWatch === null) {
          return true;
        }
        // 不对任何变动做出反应
        if (propsToWatch.length === 0) {
          return false;
        }
        // 浅比较
        return !shallowEqual(currentProps, nextProps);
      }
    }
    set(
      WithDataContainer,
      'displayName',
      `WithDataContainer(${WrappedComponent.displayName ||
        WrappedComponent.name ||
        'WrappedComponent'})`
    );
    hoistNonReactStatic(WithDataContainer, WrappedComponent);
    return WithDataContainer;
  };
}
