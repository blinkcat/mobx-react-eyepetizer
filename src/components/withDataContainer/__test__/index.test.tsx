import * as React from 'react';
import withDataContainer from '../';
import * as renderer from 'react-test-renderer';

describe('withDataContainer Hoc', () => {
  it('is function', () => {
    expect(withDataContainer).toEqual(expect.any(Function));
  });

  it('returns a function', () => {
    expect(withDataContainer(() => 1, {})).toEqual(expect.any(Function));
  });

  describe('run Hoc', () => {
    const Text = ({ text }) => {
      return <p>{text}</p>;
    };
    it('renders correctly', () => {
      const dataLoader = (props, onData) => {
        onData(null, { text: 'ok' });
      };
      const DataContainer = withDataContainer(dataLoader)(Text);
      const tree = renderer.create(<DataContainer />).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('renders Error Component', () => {
      const Error = err => {
        return <h1>{err.message}</h1>;
      };
      const dataLoader = (props, onData) => {
        onData({ message: 'have problems' });
      };
      const DataContainer = withDataContainer(dataLoader, { errorHandler: err => <Error {...err} /> })(Text);
      const tree = renderer.create(<DataContainer />).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});
