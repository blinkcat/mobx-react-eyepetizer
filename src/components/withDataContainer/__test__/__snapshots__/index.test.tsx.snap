// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withDataContainer Hoc run Hoc renders Error Component 1`] = `
<h1>
  have problems
</h1>
`;

exports[`withDataContainer Hoc run Hoc renders correctly 1`] = `
<p>
  ok
</p>
`;
